### PR TITLE
Refactor/evm sign typed data/prf 5028

### PR DESCRIPTION
**Pull request by amanCypherock**
https://app.clickup.com/t/9002019994/PRF-5028

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-5020 Port the sign message personal sign operation](https://app.clickup.com/t/9002019994/PRF-5020)

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-5028 Port the sign message typed structured data operation](https://app.clickup.com/t/9002019994/PRF-5028)

File: apps/evm_family/evm_helpers.c [hunk 1]
```diff
@@ -127,6 +129,61 @@ static inline bool is_account_hd_path(const uint32_t *path, uint32_t depth);
...
+ * null-terminated, meaning it ends with a null character
+ * ('\0'). The "value_size" parameter is the maximum number of characters to be
+ * considered from the "value
+ * @param value_size The parameter `value_size` represents the maximum size of
+ * the `value` string. It is used to ensure that the `value` string is not
+ * copied beyond its allocated memory.
+ *
+ * @return a pointer to a `ui_display_node` structure.
+ */
+static ui_display_node *evm_ui_create_display_node(const char *title,
```
Comment (ujjwal-cyph) [inline apps/evm_family/evm_helpers.c:167]: any reason to duplicate this logic? it's already in `coin_utils` and used at multiple locations.

File: apps/evm_family/evm_helpers.c [hunk 2]
```diff
@@ -223,12 +374,31 @@ bool evm_get_msg_data_hash(const evm_sign_msg_context_t *ctx, uint8_t *digest) {
...
+      result = evm_get_typed_struct_data_digest(&(ctx->typed_data), digest);
+    } break;
     default:
       break;
   }
 
   return result;
+}
+
+void evm_init_typed_data_display_node(
```
Comment (ujjwal-cyph) [inline apps/evm_family/evm_helpers.c:387]: let's keep this separate into `evm_typed_data.c` or something similar. it's not a generic helper.

File: apps/evm_family/evm_helpers.c [hunk 3]
```diff
@@ -193,6 +250,100 @@ static bool evm_get_personal_data_digest(uint8_t *msg_data,
   return true;
 }
 
+static bool evm_get_typed_struct_data_digest(
```
[the remark this answers is not in the document]
Comment (amanCypherock) [reply apps/evm_family/evm_helpers.c]: updated

File: apps/evm_family/evm_helpers.c (hunk 2 above)
Comment (amanCypherock) [reply apps/evm_family/evm_helpers.c]: updated

File: apps/evm_family/evm_helpers.c (hunk 1 above)
Comment (amanCypherock) [reply apps/evm_family/evm_helpers.c]: removed

File: apps/evm_family/evm_sign_msg.c [hunk 4]
```diff
@@ -431,11 +458,13 @@ void evm_sign_msg(evm_query_t *query) {
   }
 
   if (NULL != sign_msg_ctx.msg_data) {
+    pb_release(EVM_SIGN_TYPED_DATA_STRUCT_FIELDS, &(sign_msg_ctx.typed_data));
```
Comment (ujjwal-cyph) [inline apps/evm_family/evm_sign_msg.c:461]: this is something our of ordinary. let's add a relevant background for the requirement of this explaining the purpose and need.

Comment (ujjwal-cyph) [inline apps/evm_family/evm_sign_msg.c:461]: is the pre-condition `if (NULL != sign_msg_ctx.msg_data)` needed for calling `pb_release`?

Comment (amanCypherock) [reply apps/evm_family/evm_sign_msg.c]: updated

Comment (amanCypherock) [reply apps/evm_family/evm_sign_msg.c]: updated

File: apps/evm_family/evm_sign_msg.c [hunk 5]
```diff
@@ -200,11 +202,13 @@ static bool check_which_request(const evm_query_t *query,
 }
 
 static bool validate_initiate_query(evm_sign_msg_initiate_request_t *init_req) {
+  uint32_t size_limit = MAX_MSG_DATA_SIZE;
+
   switch (init_req->message_type) {
+    case EVM_SIGN_MSG_TYPE_SIGN_TYPED_DATA:
+      size_limit = MAX_MSG_DATA_TYPED_DATA_SIZE;
```
Comment (ujjwal-cyph) [inline apps/evm_family/evm_sign_msg.c:209]: the variable is initialized and not used in the check at line 220.

Comment (amanCypherock) [reply apps/evm_family/evm_sign_msg.c:209]: updated

**Review by ujjwal-cyph — APPROVED** [review]
LGTM

**Pull request merged**